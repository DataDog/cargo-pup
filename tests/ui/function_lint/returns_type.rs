//@compile-flags: --crate-name test_returns_type
//@compile-flags: --crate-type lib

// This test verifies that the FunctionLint's ReturnsType matcher works correctly

use std::error::Error;
use std::fmt;

// ====== Result Type Tests ======

// Function returns Result<(), i32> and should trigger the ResultErrorMustImplementError rule
// because i32 doesn't implement Error trait
fn test_result_simple() -> Result<(), i32> { //~ ERROR: Error type 'i32' in Result does not implement Error trait
    Ok(())
}

// Function returns Result with a custom error type that doesn't implement Error
fn test_result_custom_error() -> Result<String, MyError> { //~ ERROR: Error type 'MyError' in Result does not implement Error trait
    Ok("Success".to_string())
}

// ====== Option Type Tests ======

// Function returns Option<i32>
fn test_option_simple() -> Option<i32> { //~ ERROR: Function exceeds maximum length of 1 lines with 3 lines
    Some(42)
}

// Function returns Option<String>
fn test_option_string() -> Option<String> { //~ ERROR: Function exceeds maximum length of 1 lines with 3 lines
    Some("test".to_string())
}

// ====== Named Type Tests ======

// Function returns CustomType (tests Named pattern)
fn test_custom_type_function() -> CustomType { //~ ERROR: Function exceeds maximum length of 1 lines with 5 lines
    CustomType {
        value: 42
    }
}

// Function returns MyError (tests Named pattern)
fn test_my_error_function() -> MyError { //~ ERROR: Function exceeds maximum length of 1 lines with 5 lines
    MyError {
        message: "Error".to_string()
    }
}

// ====== Regex Type Tests ======

// Function returns Vec<i32> (tests Regex pattern)
fn test_vec_integers() -> Vec<i32> { //~ ERROR: Function exceeds maximum length of 1 lines with 6 lines
    let mut v = Vec::new();
    v.push(1);
    v.push(2);
    return v;
}

// Function returns Vec<String> (tests Regex pattern)
fn test_vec_strings() -> Vec<String> { //~ ERROR: Function exceeds maximum length of 1 lines with 6 lines
    let mut v = Vec::new();
    v.push("Hello".to_string());
    v.push("World".to_string());
    return v;
}

// ====== Module-based Tests ======

mod inner_module {
    // Function in module returns Result
    pub fn module_result_function() -> Result<(), i32> { 
        //~^ ERROR: Function exceeds maximum length of 2 lines with 5 lines
        //~| ERROR: Error type 'i32' in Result does not implement Error trait
        Ok(())
    }
    
    // Function in module returns Option
    pub fn module_option_function() -> Option<i32> { 
        //~^ ERROR: Function exceeds maximum length of 2 lines with 5 lines
        //~| ERROR: Function exceeds maximum length of 1 lines with 5 lines
        Some(42)
    }
}

// ====== Type Definitions ======

pub struct CustomType {
    pub value: i32,
}

// An error type that intentionally doesn't implement Error trait
pub struct MyError {
    pub message: String,
} 