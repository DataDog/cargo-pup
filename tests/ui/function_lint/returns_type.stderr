warning: Error type 'i32' in Result does not implement Error trait
  --> tests/ui/function_lint/returns_type.rs:13:1
   |
LL | fn test_result_simple() -> Result<(), i32> {
   | ^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: Consider implementing the Error trait for this type or using a type that already implements it
   = note: Applied by cargo-pup rule 'result_return_type_test'.
   = note: `#[warn(function_lint)]` on by default

warning: Error type 'MyError' in Result does not implement Error trait
  --> tests/ui/function_lint/returns_type.rs:18:1
   |
LL | fn test_result_custom_error() -> Result<String, MyError> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: Consider implementing the Error trait for this type or using a type that already implements it
   = note: Applied by cargo-pup rule 'result_return_type_test'.

warning: Function exceeds maximum length of 1 lines with 3 lines
  --> tests/ui/function_lint/returns_type.rs:25:1
   |
LL | fn test_option_simple() -> Option<i32> {
   | ^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: Consider breaking this function into smaller parts
   = note: Applied by cargo-pup rule 'option_return_type_test'.

warning: Function exceeds maximum length of 1 lines with 3 lines
  --> tests/ui/function_lint/returns_type.rs:30:1
   |
LL | fn test_option_string() -> Option<String> {
   | ^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: Consider breaking this function into smaller parts
   = note: Applied by cargo-pup rule 'option_return_type_test'.

warning: Function exceeds maximum length of 1 lines with 5 lines
  --> tests/ui/function_lint/returns_type.rs:37:1
   |
LL | fn test_custom_type_function() -> CustomType {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: Consider breaking this function into smaller parts
   = note: Applied by cargo-pup rule 'custom_type_return_test'.

warning: Function exceeds maximum length of 1 lines with 5 lines
  --> tests/ui/function_lint/returns_type.rs:44:1
   |
LL | fn test_my_error_function() -> MyError {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: Consider breaking this function into smaller parts
   = note: Applied by cargo-pup rule 'my_error_return_test'.

warning: Function exceeds maximum length of 1 lines with 6 lines
  --> tests/ui/function_lint/returns_type.rs:53:1
   |
LL | fn test_vec_integers() -> Vec<i32> {
   | ^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: Consider breaking this function into smaller parts
   = note: Applied by cargo-pup rule 'vec_return_type_test'.

warning: Function exceeds maximum length of 1 lines with 6 lines
  --> tests/ui/function_lint/returns_type.rs:61:1
   |
LL | fn test_vec_strings() -> Vec<String> {
   | ^^^^^^^^^^^^^^^^^^^^^
   |
   = help: Consider breaking this function into smaller parts
   = note: Applied by cargo-pup rule 'vec_return_type_test'.

error: Function exceeds maximum length of 2 lines with 5 lines
  --> tests/ui/function_lint/returns_type.rs:72:5
   |
LL |     pub fn module_result_function() -> Result<(), i32> { 
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: Consider breaking this function into smaller parts
   = note: Applied by cargo-pup rule 'module_based_function_length_test'.
   = note: `#[deny(function_lint)]` on by default

warning: Error type 'i32' in Result does not implement Error trait
  --> tests/ui/function_lint/returns_type.rs:72:5
   |
LL |     pub fn module_result_function() -> Result<(), i32> { 
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: Consider implementing the Error trait for this type or using a type that already implements it
   = note: Applied by cargo-pup rule 'module_result_test'.

error: Function exceeds maximum length of 2 lines with 5 lines
  --> tests/ui/function_lint/returns_type.rs:79:5
   |
LL |     pub fn module_option_function() -> Option<i32> { 
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: Consider breaking this function into smaller parts
   = note: Applied by cargo-pup rule 'module_based_function_length_test'.

warning: Function exceeds maximum length of 1 lines with 5 lines
  --> tests/ui/function_lint/returns_type.rs:79:5
   |
LL |     pub fn module_option_function() -> Option<i32> { 
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: Consider breaking this function into smaller parts
   = note: Applied by cargo-pup rule 'module_option_test'.

error: aborting due to 2 previous errors; 10 warnings emitted

